docs/proposals/dedicated-nodes.md

```diff
@@ -0,0 +1,210 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/dedicated-nodes.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Dedicated Nodes, Taints, and Tolerations
+
+## Introduction
+
+This document proposes an approach to implementing *dedicated nodes* in Kubernetes.
+A *dedicated node* is a node that can only be used by a limited set of users.
+Dedicated nodes usually occur in sets, with each set of dedicated nodes being
+associated with a particular set of users. There are many possible generalizations of dedicated
+nodes, for example a set of nodes with a particular piece of hardware could be dedicated to
+pods that require that hardware.
+
+In theory, dedicated nodes can be implemented using
+[NodeConstraint](https://github.com/kubernetes/kubernetes/pull/18261). But `NodeConstraint`
+is not well-suited for this use because `NodeConstraint` cannot easily express the idea that
+*users not in the authorized set are excluded from the dedicated nodes*, which is
+a key property of the dedicated nodes concept. Expressing
+such an exclusion rule using `NodeConstraint` would work something like the following.
+First you would label every dedicated node with `{key: "dedicated", value: "foo"}`
+where `foo` is the name of the dedicated group to whom the node is dedicated. This is
+easy. But then you would have to add to the PodSpec of every pod created by every user
+a `NodeConstraint` like `"dedicated" NotIn A,B,C,D,...` where
```

Comment:
The main benefit of taints over regular constraints is that with taints, pods that don't need to access  the "special" nodes don't need to know anything about the special nodes. So for example, if you have a cluster with no GPUs and then you add some nodes to it with GPUs and you want only pods that request GPUs to be able to schedule onto those nodes, you just add a taint to the new nodes and a toleration to the pods that request GPUs. "Regular" pods (and nodes) can be oblivious. This property becomes especially valuable once you have lots of different types of special nodes (multiple dedicated groups, multiple special hardware types, different kinds of machine problems that exclude pods, etc.). With the constraint approach, you need "regular" pods to explicitly constrain away from every different type of special node.

The one place where this becomes a little less beneficial is when there are no "regular" nodes, as in the case where you not only want to keep pods off of dedicated nodes they aren't entitled to access, but also want to keep some pods (presumably the pods of dedicated users) off of non-dedicated nodes. In that case the world is essentially as you described in your comment, where every node is effectively dedicated, and taints/tolerations are not as valuable because the non-dedicated nodes must be tainted and the non-dedicated-user pods must have tolerations. But taints/tolerations are still better than constraints even in that case because as you add dedicated node groups, you don't need to change how you handle the non-dedicated pods, whereas with the constraint approach, you would need to start adding a "dedicated != \<name of new dedicated group\>" constraint to all future and currently-pending pods.